File: algorithms/medium/numbers_with_same_consecutive_differences.ex
defmodule Solution do
  @spec nums_same_consec_diff(n :: integer, k :: integer) :: [integer]
  def nums_same_consec_diff(n, k) do
    1..9
    |> Enum.map(fn i -> {i, 1} end)
    |> :queue.from_list()
    |> bfs(n, k, [])
  end

  def bfs(q, n, k, ans) do
    if :queue.is_empty(q) do
      ans
    else
      {x, ct} = :queue.head(q)
      q = :queue.tail(q)

      if ct == n do
        bfs(q, n, k, [x | ans])
      else
        d = rem(x, 10)
        q = if d + k <= 9, do: :queue.in({x * 10 + d + k, ct + 1}, q), else: q
        q = if k != 0 && d - k >= 0, do: :queue.in({x * 10 + d - k, ct + 1}, q), else: q
        bfs(q, n, k, ans)
      end
    end
  end
end

File: algorithms/medium/numbers_with_same_consecutive_differences.rs
use std::collections::VecDeque;

impl Solution {
    pub fn nums_same_consec_diff(n: i32, k: i32) -> Vec<i32> {
        let mut q = VecDeque::new();
        for i in 1..=9 {
            q.push_back((i, 1));
        }
        let mut ans = Vec::new();
        while !q.is_empty() {
            let (x, ct) = q.pop_front().unwrap();
            if ct == n {
                ans.push(x);
                continue;
            }
            let d = x % 10;
            if d + k <= 9 {
                q.push_back((x * 10 + d + k, ct + 1));
            }
            if d - k >= 0 && k != 0 {
                q.push_back((x * 10 + d - k, ct + 1));
            }
        }
        ans
    }
}
